### PR TITLE
Implement new high-level command API

### PR DESCRIPTION
**Pull request by dualspiral**
[Sponge API](https://github.com/SpongePowered/SpongeAPI/pull/1587) | **SpongeCommon** | [SpongeForge](https://github.com/SpongePowered/SpongeForge/pull/1667)

This now works - see API PR for more description.

### CODE REVIEW

**Comment by michaelzangl** [comment]
Just if you need it: For Minebot I needed a lot of [commands](https://github.com/michaelzangl/minebot/tree/forge-1.12/Minebot/src/net/famzangl/minecraft/minebot/ai/commands). This is why I wrote my own [command parser](https://github.com/michaelzangl/minebot/tree/forge-1.12/Minebot/src/net/famzangl/minecraft/minebot/ai/command). Parsing is done by the `CommandDefinition` class. Commands can simply be defined in a class:

	@AICommand(name = "minebuild", helpText = "Fill an area with a given block.")
	public class CommandFillArea {
		public static AIStrategy run(
			AIHelper helper,
			@AICommandParameter(type = ParameterType.FIXED, fixedName = "fill", description = "") String nameArg,
			@AICommandParameter(type = ParameterType.BLOCK_NAME, blockFilter=FillAreaPathfinder.FillBlocks.class, description = "The block to place") BlockWithDataOrDontcare block) {
			...
		}
	}

This automatically adds the help text for the command with the parameter description. Tab complete is fully supported. Parameter types can be blocks, enums, strings, numbers, BlockPos. It is could be made easy to add a new parameter type by using a class instead of an enum constant.
Command invocations are then:
`/minebuild fill stone`

(currently licensed under GPL, but I have no problem with re-licensing my parts under MIT)

**Comment by gabizou** [comment]
If you can, when you update the implementation with bleeding, provide a test plugin to make some examples of how things would look with the API changes. I think it would suit better with the description overall.

**Comment by dualspiral** [comment]
Okie doke, will do.

**Comment by ryantheleach** [comment]
This enables https://github.com/SpongePowered/SpongeCommon/issues/1463 right?

**Comment by dualspiral** [comment]
Yes.

**Comment by parlough** [comment]
@dualspiral Making all of our dreams come true.

**Comment by zml2008** [comment]
it's finally happening!

File: src/main/java/co/aikar/timings/TimingsExport.java [hunk 1]
```diff
@@ -116,7 +116,7 @@ static void reportTimings(CommandSource sender) {
                 .add("maxmem", runtime.maxMemory())
                 .add("cpu", runtime.availableProcessors())
                 .add("runtime", ManagementFactory.getRuntimeMXBean().getUptime())
-                .add("flags", RUNTIME_FLAG_JOINER.join(runtimeBean.getInputArguments()))
+                .add("setFlags", RUNTIME_FLAG_JOINER.join(runtimeBean.getInputArguments()))
```
Comment (kashike) [inline src/main/java/co/aikar/timings/TimingsExport.java]: no

**Review by kashike — CHANGES_REQUESTED** [review]

File: src/main/java/org/spongepowered/common/command/parameter/flag/SpongeFlags.java [hunk 2]
```diff
@@ -0,0 +1,210 @@
...
+            return this.args.createError(message, inner);
+        }
+
+        @Override
+        public String rawArgsFromCurrentPosition() {
+            return this.args.rawArgsFromCurrentPosition();
+        }
+
+        private ArgumentParseException createValueError() {
+            return createError(t("Short setFlags that are not at the end of a group cannot have a value."));
```
Comment (kashike) [inline src/main/java/org/spongepowered/common/command/parameter/flag/SpongeFlags.java]: Should probably be flags, not setFlags

Comment (dualspiral) [reply src/main/java/org/spongepowered/common/command/parameter/flag/SpongeFlags.java]: I'll get this right one day...

**Comment by dualspiral** [comment]
Closed in favour of https://github.com/SpongePowered/SpongeAPI/pull/1958 for 1.13

**Pull request closed without merging**